### PR TITLE
Support #68, #79, #80, #83 and fixes #60, #78

### LINKED ISSUE #60: Komatsu Forwarder Pack (HR Forst und Fahrzeugbau)
Now i don't know what i'm doing wrong but in version 1.1.0.1 it says this is included. https://www.farming-simulator.com/mod.php?mod_id=247034&title=fs2022  have tried in all different lengths 

//MrSaffron

### PR DESCRIPTION
**Pull request by KingArthurLV**
Added support for:
- [FS22_AmericanFlatbedPack](https://www.farming-simulator.com/mod.php?mod_id=264240&title=fs2022)
- [FS22_RollandRPLCH](https://www.farming-simulator.com/mod.php?mod_id=240058&title=fs2022)
- [FS22_AW362_CaboverR](https://www.farming-simulator.com/mod.php?mod_id=250565&title=fs2022)
- [FS22_LizardF600](https://www.farming-simulator.com/mod.php?mod_id=243774&title=fs2022)

Fixes for:
- [FS22_LizardLT679SmallLogTrailer](https://www.farming-simulator.com/mod.php?mod_id=245565&title=fs2022)
- [FS22_Komatsu_Forwarder_Pack](https://www.farming-simulator.com/mod.php?mod_id=247034&title=fs2022)

Everything tested with various sizes of palets, bales and wood logs.
In testing used only FS22_UniversalAutoload v.1.3.3.0 and FS22_EasyDevControls v.1.3.0.2

### CODE REVIEW

**Comment by KingArthurLV** [comment]
I have quastion - if I use isBaleTrailer="true" does enableSideLoading="true" enableRearLoading="true" are affecting it or it is redundant?

**Comment by ddewar** [comment]
Bale trailer and side/rear loading are different features. Both should be declared. If both are used.

**Comment by ddewar** [comment]
Thank you very much for this PR. Looks like good stuff. I'm going to hold off merging it for a few days here. We have an updated UALAddon version with Giants right now. So just waiting to see what they do with that before pushing more vehicles to the repo.

**Comment by KingArthurLV** [comment]
Added support for:
- [FS22_ITRTwo](https://www.farming-simulator.com/mod.php?mod_id=259967&title=fs2022)

Fix for:
- [FS22_Bailey_Bale_Pallet_Trailer](https://www.farming-simulator.com/mod.php?mod_id=228030&title=fs2022)

**Comment by KingArthurLV** [comment]
Added support for:
- [Marshall BC/32](https://www.farming-simulator.com/mod.php?mod_id=235656&title=fs2022)
- [Junkkari J16-J18](https://www.farming-simulator.com/mod.php?mod_id=249628&title=fs2022)

**Comment by KingArthurLV** [comment]
Added support:
- [FS22_Old20FootBaleTrailer](https://www.farming-simulator.com/mod.php?mod_id=267221&title=fs2022)
- [FS22_manTGX_26640_strawbale](https://www.farming-simulator.com/mod.php?mod_id=264434&title=fs2022)
- [FS22_lizard_rp03](https://www.farming-simulator.com/mod.php?mod_id=264436&title=fs2022)
- [FS22_camara_rp03](https://www.farming-simulator.com/mod.php?mod_id=264435&title=fs2022)

Fix for:
- [FS22_lizard_sp03_crossplay](https://www.farming-simulator.com/mod.php?mod_id=243958&title=fs2022)
- [FS22_camara_sp03](https://www.farming-simulator.com/mod.php?mod_id=243957&title=fs2022)

**Comment by ddewar** [comment]
Hey there, do me a favor and hold off on adding anything else to this PR. I've started checking configs and testing these. Let's get these in and if you do anymore let's get them in a new PR.

**Comment by KingArthurLV** [comment]
> Hey there, do me a favor and hold off on adding anything else to this PR. I've started checking configs and testing these. Let's get these in and if you do anymore let's get them in a new PR.

Sure man! No problem. Let me know if there is something I need to fix or adjust in this PR.

Also I would like to purpose to create one thing in this repo - guidlines for creating support. Because what I noticed, that different people have different opinions on loading sizes, especialy height. I took facts as my guide. I studied what are real world parameters for different loads and tok it as general rule. As well I allowed some small adjusments to acomodate maximum reasonable amount of items in loading.

**Comment by ddewar** [comment]
Looking good so far. First thing I've run into, maybe we shouldn't support UAL on the F600 that has a capacity?

image

Also, width on that configuration is a little high. Pallets are getting suck in the tipper sides.

**Comment by KingArthurLV** [comment]
Hi! I mean, I also thought about it, but at the same time, that is tipper with low boards, so in real world situation, truck would be used as pallet hauler as well. But I also understand that because of game limitations, people would be able lead bulk alongside pallets.
My personal opinion is giving players maximum real-world gameplay and give freedom to choose what to load in the truck using convenience of Autoload.

About width, I will check that out, I believe, I hoped that flatbed and low board tipper was with same width. Probably will need to make additional support line for that.

**Comment by KingArthurLV** [comment]
fsScreen_2023_05_10_21_40_30

I checked it, so "pallet" itself is not clipping into tipper boards, but because each pallet has extra 0.5 (dont know exact number) xyz border dimension, loading size needs to be a bit wider, to actually put 2 pallets horizontally, to maximize pallets capacity.

I actually would love to know, why exactly all items have extra blank space. To not cause collisions?

**Comment by KingArthurLV** [comment]
And maybe we can make suggestion to FS22_UniversalAutoload mod to create additional check to prevent player to load pallets if tipper for example is filled?

**Comment by ddewar** [comment]
> fsScreen_2023_05_10_21_40_30
> 
> I checked it, so "pallet" itself is not clipping into tipper boards, but because each pallet has extra 0.5 (dont know exact number) xyz border dimension, loading size needs to be a bit wider, to actually put 2 pallets horizontally, to maximize pallets capacity.
> 
> I actually would love to know, why exactly all items have extra blank space. To not cause collisions?

Yeah, I think with a lot of base game pallets Loki left a little wiggle room to avoid clipping. I'm not sure that will always be the case for all modded pallets though. I try to use the outlines as the strict rules, even if the pallets themselves do not clip.

**Comment by ddewar** [comment]
> And maybe we can make suggestion to FS22_UniversalAutoload mod to create additional check to prevent player to load pallets if tipper for example is filled?

Not a bad idea, but the problem really comes in when you autoload something like a pallet of seeds. When it gets put onto the trailer it starts unloading and that's when it starts causing problems. I'm not against leaving it and letting the user deal with it if they try loading non-production pallets, but for the most part I have avoided trailers with capacity to avoid that issue. Especially, if the trailer has another version that can be without capacity like in this case.

**Comment by ddewar** [comment]
I'm a little concerned about the length on the Marshall trailer. I'm running into some colliding issues. 

image


Any reason you wanted the longer length?

**Comment by ddewar** [comment]
image

Some funkiness with the Junkkari. But I think it's good enough. Run into any issues with it?

**Comment by ddewar** [comment]
The other flatbed on the Junkkari looks like it needs a different length and offset. It clips through the front when added and goes too far off the back.

image

**Comment by ddewar** [comment]
image

TGX Man Straw Bale mod might need two things:

1. Only limit it to the config without autoload
2. The loading zone needs to be moved back just a tiny bit. Some pallets are getting stuck and flying all over the place

**Comment by KingArthurLV** [comment]
Nice, Dan is doing some testing :)

Okey first about bale height and height overall. I dig up 2 numbers from real world truckload height in US and EU. In EU it is 4,5m and in US 4,3m (14 feet) including trailer height from the ground.
I took bale height 3,8m because it is minimum height to allow load 2 layers of 1,8m bales.
For pallets I am thinking to use max 3 layers of highest pallets which is 1,3m (wool) and 1,2m (cereal, flour, sugar).
At the same time, I believe it would make sense to also use correct proportions. Let’s say trailer length is less than 1,3 x 3 = 3,9m, so then height should not be more than trailer height.

Basically, this is what I wrote before, that I think it would really help, if we could come up with some exact guidelines and make it as standard.

**Comment by KingArthurLV** [comment]
> image
> 
> Some funkiness with the Junkkari. But I think it's good enough. Run into any issues with it?

I tested it with different sizes of pallets and did not have an issue.
But I have an idea, that maybe we can make suggestion to FS22_UniversalAutoload mod, to have option to create different loading areas for different container type. Because that would really make sense and we will be able to really make custom loading spaces for bales and pallets separate. I know it would create much more work in our end, but it would make much pleasant gaming experience and use of FS22_UniversalAutoload.

And I understand that I am marching with all these ideas and changes, and You don’t know me and my skills and need to verify if I am legit :). That is why I am making an effort to it. Secondly, I really like this game and the community around it, but also, I see so much potential and some "holes" that can be improved, patched etc.

**Comment by KingArthurLV** [comment]
> > And maybe we can make suggestion to FS22_UniversalAutoload mod to create additional check to prevent player to load pallets if tipper for example is filled?
> 
> Not a bad idea, but the problem really comes in when you autoload something like a pallet of seeds. When it gets put onto the trailer it starts unloading and that's when it starts causing problems. I'm not against leaving it and letting the user deal with it if they try loading non-production pallets, but for the most part I have avoided trailers with capacity to avoid that issue. Especially, if the trailer has another version that can be without capacity like in this case.

Okey, I understand and that makes sense as well. So, I will put it as a guideline in my draft that I am preparing. General rule - no autoload for bulk item tippers, so there is no confusion.

**Comment by KingArthurLV** [comment]
> image
> 
> TGX Man Straw Bale mod might need two things:
> 
> 1. Only limit it to the config without autoload
> 2. The loading zone needs to be moved back just a tiny bit. Some pallets are getting stuck and flying all over the place

I see that I need to improve my testing ground, most of the time I load same material type pallets at once, not random. I will retest all, to make sure that there are no funky things happening.

**Comment by ddewar** [comment]
> > image
> > TGX Man Straw Bale mod might need two things:
> > 
> > 1. Only limit it to the config without autoload
> > 2. The loading zone needs to be moved back just a tiny bit. Some pallets are getting stuck and flying all over the place
> 
> I see that I need to improve my testing ground, most of the time I load same material type pallets at once, not random. I will retest all, to make sure that there are no funky things happening.

The ualAddPallets command will do the randomization for you. Super helpful!

**Comment by ddewar** [comment]
> > image
> > Some funkiness with the Junkkari. But I think it's good enough. Run into any issues with it?
> 
> I tested it with different sizes of pallets and did not have an issue. But I have an idea, that maybe we can make suggestion to FS22_UniversalAutoload mod, to have option to create different loading areas for different container type. Because that would really make sense and we will be able to really make custom loading spaces for bales and pallets separate. I know it would create much more work in our end, but it would make much pleasant gaming experience and use of FS22_UniversalAutoload.
> 
> And I understand that I am marching with all these ideas and changes, and You don’t know me and my skills and need to verify if I am legit :). That is why I am making an effort to it. Secondly, I really like this game and the community around it, but also, I see so much potential and some "holes" that can be improved, patched etc.

All good, you're making some good suggestions. Loki and I just haven't had much of a need for too many advanced features without making it really tough to manage as a user. I don't think we need zones for bales vs pallets in these situations. The big thing is just adjusting the zones on the trailers to always match the planes that are being loaded. I know sometimes, especially with bales, you can get away with oversizing them, but most of the time it just comes back to bite us when using pallets. The only time I try to do oversizing is when it is a log trailer and we can get to that next full meter mark.

As far as height goes, I don't mind your suggestion on 3.8 for baleHeights. I find it very frustrating to not be able to stack 2 180 round bales, but when it comes to pallets, I get lots of issues on these bigger heights. Remembering that by default UAL will stack starting at the front up to max height, sometimes this results in very top-forward heavy loading. So, for the most part, I have been making trailers support about 2 high for the pallets. Anything over that was getting very tippy. I always encourage users who want more height to copy the default config into their local UAL settings and increase the height. 

My suggestion would be to check your loading zones for:

1. Height causing high pallet stacking (Bales are okay)
2. The loading zones should not extend past the loading bed.

To me it looked like these conifgs are all 90% there and I really appreciate the work you've put into them. Honestly, might need to start using some of these trailers in my own save games :)

**Comment by ddewar** [comment]
> > > And maybe we can make suggestion to FS22_UniversalAutoload mod to create additional check to prevent player to load pallets if tipper for example is filled?
> > 
> > 
> > Not a bad idea, but the problem really comes in when you autoload something like a pallet of seeds. When it gets put onto the trailer it starts unloading and that's when it starts causing problems. I'm not against leaving it and letting the user deal with it if they try loading non-production pallets, but for the most part I have avoided trailers with capacity to avoid that issue. Especially, if the trailer has another version that can be without capacity like in this case.
> 
> Okey, I understand and that makes sense as well. So, I will put it as a guideline in my draft that I am preparing. General rule - no autoload for bulk item tippers, so there is no confusion.
> 
> BUT if could make suggestion to Loki create one more parameter for each loading area to set containerType which is laready defined in ContainerTypes.xml, we could still use this tipper, just limit it for bales and/or liquid tanks etc.

Good rule of thumb there. 

Maybe worth opening a suggestion on the UAL GitHub for how to handle bulk trailers. I agree, they can make things messy if you're not careful as is!

**Comment by KingArthurLV** [comment]
I am retesting all commits, I think today will do it all.

I made also suggestion to UAL: https://github.com/loki79uk/FS22_UniversalAutoload/issues/177

**Comment by ddewar** [comment]
> I am retesting all commits, I think today will do it all.
> 
> I made also suggestion to UAL: [loki79uk/FS22_UniversalAutoload#177](https://github.com/loki79uk/FS22_UniversalAutoload/issues/177)

How we doing on this? Am I good to look at it again?

**Comment by KingArthurLV** [comment]
Iam half way trough. Made some adjustmwnts. Unfortunatley I was 2 days out of town, now I am back. Will try today/tomorrow finish it.
Is it okey if it will be one commit for all adjustments?

**Comment by ddewar** [comment]
> Iam half way trough. Made some adjustmwnts. Unfortunatley I was 2 days out of town, now I am back. Will try today/tomorrow finish it. Is it okey if it will be one commit for all adjustments?

All good here with one commit. No rush, just ping me whenever I should look again.

**Comment by KingArthurLV** [comment]
Hi! Finaly got some free time from IRL to finish up testing and fixing issues. Check it out @ddewar  ;)

**Comment by ddewar** [comment]
Trying to finish this review up tonight. Please do me a favor and remove any changes to supportedMods.md when you get a chance. It's auto generated, and we won't worry about it.

**Comment by KingArthurLV** [comment]
> 

Ou sorry about that, did not know that. I will do that.

**Comment by ddewar** [comment]
> > 
> 
> Ou sorry about that, did not know that. I will do that. You need me to remove all changes I made in that file, or just changes that I made in secondary testing?

The file completely. Just need the supported vehicles file.

**Comment by ddewar** [comment]
Hey, just letting you know, I'm about halfway through and these are looking pretty good. The lower stack heights for pallets is making it a little more manageable. We'll have to see if we get any complaints with the 3 high of almost every pallet, but looks good to me.

**Comment by ddewar** [comment]
Okay, just about there, everything is looking great. If you can, test the Marshall trailer with 125cm round bales. I think there's a collision on the back we are hitting. Might have to shorten that one just a touch.

**Comment by KingArthurLV** [comment]
Just checked, are You sure problem is with 125 cm? For me seems perfect.
fsScreen_2023_05_24_21_51_19

**Comment by ddewar** [comment]
> Just checked, are You sure problem is with 125 cm? For me seems perfect. fsScreen_2023_05_24_21_51_19

Ah, try stack loading. My tests run in stack first.

**Comment by ddewar** [comment]
> Just checked, are You sure problem is with 125 cm? For me seems perfect. fsScreen_2023_05_24_21_51_19

PS. I love how that trailer looks with horizonal loading.

**Comment by KingArthurLV** [comment]
Will do, I actually planning to go through all Supported list and check if there is anything to improve or optimize. I am an efficiency freak and love when everything is tidy and standardized.

BTW what is Your background? Programmer? What languafes You work with?

**Comment by ddewar** [comment]
One last question. Looks like on the Komatsu pack you removed the dynamic height axis? Was it note working anymore or just a preference?

```            <loadingArea offset="0 1.40 -3.45" width="2.70" height="0.6" length="5.0" heightAxis="AXIS_CRANE_TOOL3" noLoadingIfUnfolded="true"/>
```

**Comment by ddewar** [comment]
> 

Happy to discuss standardizing and such. If you have a Discord reach out there, might be a little better. 

Discord: ddewar#1716

**Comment by KingArthurLV** [comment]
> 

Actually not remmber anymore, but I remmeber that trailer was messy to load. Will recheck. and make fixes if needed.

**Review by ddewar — APPROVED** [review]
Let's get this in. Great stuff.

File: supportedMods.md
[the remark this answers is not in the document]
Comment (KingArthurLV) [reply supportedMods.md:1]: I did it by hand.
